### PR TITLE
Add module for reverse DNS on exoscale VMs

### PR DESCRIPTION
**Pull request by znerol**
Migrated from PR ansible/ansible#57319

### CODE REVIEW

**Comment by znerol** [comment]
What is the significance of `version_added` metadata field with this new project structure?

**Comment by znerol** [comment]
From the sanity test log:
```
2020-08-18T13:57:00.6559686Z ERROR: Command "/usr/bin/python3.6 /root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate-modules --format json --arg-spec plugins/modules/exo_dns_domain.py plugins/modules/exo_dns_record.py plugins/modules/exo_rdns_record.py --collection ansible_collections/ngine_io/exoscale --collection-version 1.0.0" returned exit status 1.
2020-08-18T13:57:00.6560089Z >>> Standard Error
2020-08-18T13:57:00.6560221Z Traceback (most recent call last):
2020-08-18T13:57:00.6562615Z   File "/root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate-modules", line 8, in <module>
2020-08-18T13:57:00.6562880Z     main()
2020-08-18T13:57:00.6563554Z   File "/root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 2444, in main
2020-08-18T13:57:00.6563730Z     run()
2020-08-18T13:57:00.6564171Z   File "/root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 2341, in run
2020-08-18T13:57:00.6564428Z     mv1.validate()
2020-08-18T13:57:00.6564869Z   File "/root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 2145, in validate
2020-08-18T13:57:00.6565041Z     doc_info, docs = self._validate_docs()
2020-08-18T13:57:00.6565473Z   File "/root/ansible/test/lib/ansible_test/_data/sanity/validate-modules/validate_modules/main.py", line 986, in _validate_docs
2020-08-18T13:57:00.6565917Z     add_fragments(doc, self.object_path, fragment_loader=fragment_loader, is_module=True)
2020-08-18T13:57:00.6566126Z   File "/root/ansible/lib/ansible/utils/plugin_docs.py", line 198, in add_fragments
2020-08-18T13:57:00.6566838Z     raise AnsibleError('unknown doc_fragment(s) in file {0}: {1}'.format(filename, to_native(', '.join(unknown_fragments))))
2020-08-18T13:57:00.6567235Z ansible.errors.AnsibleError: unknown doc_fragment(s) in file plugins/modules/exo_rdns_record.py: cloudstack
```

File: plugins/modules/exo_rdns_record.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+# -*- coding: utf-8 -*-
+#
+# (c) 2019, Lorenz Schori <lo@znerol.ch>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:11]: ANSIBLE_METADATA block can be removed

File: plugins/modules/exo_rdns_record.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    description:
+      - Reverse DSN name for the cs instance. Required if state=present.
+    type: str
+  state:
+    description:
+      - State of the record.
+    default: present
+    choices: [ 'present', 'absent' ]
+    type: str
+extends_documentation_fragment: cloudstack
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:41]: this adds a new dependecy to the cloudstack collection, please change it to

`ngine_io.cloudstack.cloudstack`

File: plugins/modules/exo_rdns_record.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+'''
+
+EXAMPLES = '''
+- name: Set the reverse DNS for a virtual machine
+  local_action:
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:46]: this doesn't look right, should look like
```yaml
- name: Set the reverse DNS for a virtual machine
  ngine_io.exoscale.exo_rdns_record:
    name: web-vm-1
    content: www.example.com
```

File: plugins/modules/exo_rdns_record.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+---
+exo_rdns_domain:
+  description: Reverse DSN name for the cs instance
+  returned: success
+  type: str
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_bytes, to_text
+from ansible.module_utils.cloudstack import (
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:66]: please change it to

`ansible_collections.ngine_io.cloudstack.plugins.module_utils.cloudstack import (`

File: plugins/modules/exo_rdns_record.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: exo_rdns_record
+short_description: Manages reverse DNS records for Exoscale compute instances.
+description:
+    - Set and unset reverse DNS record on Exoscale instance.
+version_added: "2.10"
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:22]: version_added shoud be next minior in semver -> `1.1.0`

File: galaxy.yml [hunk 6]
```diff
@@ -1,7 +1,8 @@
 ---
 authors:
   - "René Moser <mail@renemoser.net>"
-dependencies: {}
+dependencies:
+  - ngine_io.cloudstack
```
Comment (resmo) [inline galaxy.yml:5]:
~~~yaml
dependencies:
  ngine_io.cloudstack: '>=1.0.0'
~~~

**Comment by resmo** [comment]
adding a collection dependency is also new to me, I'll try to find out how this works ;)

**Comment by znerol** [comment]
Does the sanity workflow actually install dependencies? Is this a responsibility of `ansible-test` when preparing the docker images?

**Comment by znerol** [comment]
[ansible-collections/community.aws](/ansible-collections/community.aws) has some dependencies, but they pull them in using a [mix of git and ansible-galaxy commands](https://github.com/ansible-collections/community.aws/blob/1a1be5f494bb3935084b33d9183a41b9e9389cff/tests/utils/shippable/shippable.sh#L80-L93).

**Comment by znerol** [comment]
Sanity is back green. Regrettably I cannot test this against a real exoscale instance since I currently do not maintain any infra running there.

**Comment by resmo** [comment]
@znerol perfect, I have an account where I can run tests, should have time today to do so. I'll let you know

File: plugins/modules/exo_rdns_record.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+options:
+  name:
+    description:
+      - Name of the compute instance
+    required: true
+    type: str
+  content:
+    aliases: [ 'value' ]
+    description:
+      - Reverse DSN name for the cs instance. Required if state=present.
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:28]:
```diff
-      - Reverse DSN name for the cs instance. Required if state=present.
+      - Reverse DNS name for the compute instance. Required if C(state=present).
```

File: plugins/modules/exo_rdns_record.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+        self.instance = None
+
+    def get_instance(self):
+        instance = self.instance
+        if not instance:
+            args = {
+                'account': self.get_account(key='name'),
+                'domainid': self.get_domain(key='id'),
+                'projectid': self.get_project(key='id'),
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:88]: Exoscale's API does not know these 3 params: https://community.exoscale.com/api/compute/#listvirtualmachines_GET and that is why we should remove them here.

File: plugins/modules/exo_rdns_record.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                }
+
+        return result
+
+    def present_record(self):
+        instance = self.get_instance()
+
+        if not instance:
+            self.module.fail_json(
+                msg="No virtual machine with name=%s found. " % self.name)
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:128]: for consistency
```diff
-                msg="No virtual machine with name=%s found. " % self.name)
+                msg="No compute instance with name=%s found. " % self.name)
```

File: plugins/modules/exo_rdns_record.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+        if self.has_changed(data, record):
+            self.result['changed'] = True
+            self.result['diff']['before'] = record
+            self.result['diff']['after'] = data
+            if not self.module.check_mode:
+                self.query_api('updateReverseDnsForVirtualMachine',
+                               id=instance['id'],
+                               domainname=data['domainname'])
+
+        return data
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:143]: Rather then returning the user input, I would prefer to return what the api returns in the result of [updateReverseDnsForVirtualMachine](https://community.exoscale.com/api/compute/#updatereversednsforvirtualmachine_GET)

File: plugins/modules/exo_rdns_record.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+- name: Delete the reverse DNS for a virtual machine
+  ngine_io.exoscale.exo_rdns_record:
+    name: web-vm-1
+    state: absent
+'''
+
+RETURN = '''
+---
+exo_rdns_domain:
+  description: Reverse DSN name for the cs instance
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:54]:
```diff
-   description: Reverse DSN name for the cs instance
+   description: Reverse DNS name of the compute instance
```

**Review by resmo — CHANGES_REQUESTED** [review]

**Comment by resmo** [comment]
I just noticed another thing to fix: there is API for adding reverse dns for public IPs: queryReverseDnsForPublicIpAddress: we should rename this module to not collide with the names. suggestion: `instance_rdns_record`?

(I'd like also to get rid of the `exo_` prefix for new and change existing in the long term because we have namespacing built in in collections)

File: plugins/modules/exo_rdns_record.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+# (c) 2019, Lorenz Schori <lo@znerol.ch>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: exo_rdns_record
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:13]: rename to `instance_rdns_record`?

File: plugins/modules/exo_rdns_record.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+      - State of the record.
+    default: present
+    choices: [ 'present', 'absent' ]
+    type: str
+extends_documentation_fragment: ngine_io.cloudstack.cloudstack
+'''
+
+EXAMPLES = '''
+- name: Set the reverse DNS for a virtual machine
+  ngine_io.exoscale.exo_rdns_record:
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:41]: rename to `instance_rdns_record`?

File: plugins/modules/exo_rdns_record.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    name: web-vm-1
+    content: www.example.com
+
+- name: Delete the reverse DNS for a virtual machine
+  ngine_io.exoscale.exo_rdns_record:
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:46]: rename to `instance_rdns_record`?

File: plugins/modules/exo_rdns_record.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    name: web-vm-1
+    state: absent
+'''
+
+RETURN = '''
+---
+exo_rdns_domain:
```
Comment (resmo) [inline plugins/modules/exo_rdns_record.py:53]: rename to `instance_rdns_domain`?

**Review by resmo — CHANGES_REQUESTED** [review]

File: plugins/modules/exo_rdns_record.py (hunk 10 above)
Comment (znerol) [reply plugins/modules/exo_rdns_record.py]: Regrettably I can hardly help with that. I currently neither have access to exoscale nor do I have a use case for the return values. I'd rather prefer to leave that up to the maintainers / active users of this collection.

File: plugins/modules/instance_rdns_record.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        if self.has_changed(data, record):
+            self.result['changed'] = True
+            self.result['diff']['before'] = record
+            self.result['diff']['after'] = data
+            if not self.module.check_mode:
+                self.query_api('updateReverseDnsForVirtualMachine',
+                               id=instance['id'],
+                               domainname=data['domainname'])
+
+        return data
```
Comment (resmo) [inline plugins/modules/instance_rdns_record.py:140]: I would return what we get from the api.
~~~diff
-                self.query_api('updateReverseDnsForVirtualMachine',
-                              id=instance['id'],
-                              domainname=data['domainname'])

-        return data

+                record = self.query_api(
+                    'updateReverseDnsForVirtualMachine',
+                    id=instance['id'],
+                    domainname=data['domainname']
+                )
+                if 'id' in record: del record['id']

+        return record
~~~

**Comment by resmo** [comment]
Thanks for your efforts!

**Pull request merged**